icicle/tests/test_hash_api.cpp

```diff
@@ -0,0 +1,215 @@
+  // Run single hash operation
+  ICICLE_CHECK(sha3_256.hash_single(input.get(), output.get(), config));
+  // TODO: Verify output (e.g., check CPU against CUDA)
+}
+
+TEST_F(HashApiTest, sha3_512)
+{
+  const uint64_t nof_input_limbs = 16; // Number of input limbs
+  // Create unique pointers for input and output arrays
+  auto input = std::make_unique<uint32_t[]>(nof_input_limbs);
+  auto output = std::make_unique<uint32_t[]>(nof_input_limbs);
+  // Randomize the input array
+  randomize(input.get(), nof_input_limbs);
+
+  auto config = default_hash_config();
+  // Create sha3-512 hash object
+  auto sha3_512 = create_sha3_512_hash(nof_input_limbs);
+  // Run single hash operation
+  ICICLE_CHECK(sha3_512.hash_single(input.get(), output.get(), config));
+  // TODO: Verify output (e.g., check CPU against CUDA)
+}
+
+/****************************** Merkle **********************************/
+class HashSumBackend : public HashBackend
+{
+public:
+  HashSumBackend(uint64_t total_input_limbs, uint64_t total_output_limbs, uint64_t total_secondary_input_limbs = 0)
+      : HashBackend(total_input_limbs, total_output_limbs, total_secondary_input_limbs)
+  {
+  }
+
+  virtual eIcicleError hash_single(
+    const limb_t* input_limbs,
+    limb_t* output_limbs,
+    const HashConfig& config,
+    const limb_t* secondary_input_limbs = nullptr) const
+  {
+    output_limbs[0] = 0;
+    for (int i = 0; i < m_total_input_limbs; ++i) {
+      output_limbs[0] += input_limbs[i];
+    }
+    for (int i = 1; i < m_total_output_limbs; ++i) {
+      output_limbs[i] += output_limbs[0];
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  virtual eIcicleError hash_many(
+    const limb_t* input_limbs,
+    limb_t* output_limbs,
+    int nof_hashes,
+    const HashConfig& config,
+    const limb_t* secondary_input_limbs = nullptr) const
+  {
+    for (int i = 0; i < nof_hashes; ++i) {
+      hash_single(input_limbs, output_limbs, config, secondary_input_limbs);
+      input_limbs += m_total_input_limbs;
+      output_limbs += m_total_output_limbs;
+      secondary_input_limbs += m_total_secondary_input_limbs;
+    }
+    return eIcicleError::SUCCESS;
+  }
+};
+
+Hash create_hash_sum(uint64_t total_input_limbs, uint64_t total_output_limbs, uint64_t total_secondary_input_limbs = 0)
+{
+  auto backend = std::make_shared<HashSumBackend>(total_input_limbs, total_output_limbs, total_secondary_input_limbs);
+  return Hash(backend);
+}
+
+TEST_F(HashApiTest, MerkleTree)
+{
+  // const uint64_t nof_leaves_limbs = 100; // Number of input limbs
+  // auto leaves = std::make_unique<uint32_t[]>(nof_leaves_limbs);
+  // // Randomize the input array
+  // randomize(leaves.get(), nof_leaves_limbs);
+
+  auto config = default_merkle_tree_config();
+  auto layer0_hash = create_hash_sum(5, 2); // input 5 limbs, output 2 limbs
+  auto layer1_hash = create_hash_sum(6, 3); // input 6 limbs, output 3 limbs
+  auto layer2_hash = create_hash_sum(6, 2); // input 6 limbs, output 2 limbs
```

Comment:
What about testing hashes with side inputs?

Reply:
Don't worry you will add all the tests. I am just doing the API